lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -17,18 +32,75 @@ const SCHEDULABLE_TASK_TITLE_ALT2 = 'ThreadControllerImpl::DoWork';
 // m65 and earlier
 const SCHEDULABLE_TASK_TITLE_ALT3 = 'TaskQueueManager::ProcessTaskFromWorkQueue';
 
+class TraceProcessor {
+  /**
+   * @return {Error}
+   */
+  static createNoNavstartError() {
+    return new Error('No navigationStart event found');
+  }
 
-const LHError = require('../lh-error.js');
+  /**
+   * @return {Error}
+   */
+  static createNoFirstContentfulPaintError() {
+    return new Error('No firstContentfulPaint event found');
+  }
+
+  /**
+   * @return {Error}
+   */
+  static createNoTracingStartedError() {
+    return new Error('No tracingStartedInBrowser event found');
+  }
+
+  /**
+   * Returns true if the event is a navigation start event of a document whose URL seems valid.
+   *
+   * @param {LH.TraceEvent} event
+   */
+  static _isNavigationStartOfInterest(event) {
+    return event.name === 'navigationStart' &&
+      (!event.args.data || !event.args.data.documentLoaderURL ||
+        ACCEPTABLE_NAVIGATION_URL_REGEX.test(event.args.data.documentLoaderURL));
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} traceEvents
+   * @param {(e: LH.TraceEvent) => boolean} filter
+   */
+  static _filteredStableSort(traceEvents, filter) {
+    // create an array of the indices that we want to keep
+    const indices = [];
+    for (let srcIndex = 0; srcIndex < traceEvents.length; srcIndex++) {
+      if (filter(traceEvents[srcIndex])) {
+        indices.push(srcIndex);
+      }
+    }
+
+    // sort by ts, if there's no ts difference sort by index
+    indices.sort((indexA, indexB) => {
+      const result = traceEvents[indexA].ts - traceEvents[indexB].ts;
+      return result ? result : indexA - indexB;
+    });
+
+    // create a new array using the target indices from previous sort step
+    const sorted = [];
+    for (let i = 0; i < indices.length; i++) {
+      sorted.push(traceEvents[indices[i]]);
+    }
+
+    return sorted;
+  }
 
-class TraceProcessor {
   /**
    * There should *always* be at least one top level event, having 0 typically means something is
    * drastically wrong with the trace and we should just give up early and loudly.
    *
    * @param {LH.TraceEvent[]} events
    */
   static assertHasToplevelEvents(events) {
-    const hasToplevelTask = events.some(TraceProcessor.isScheduleableTask);
+    const hasToplevelTask = events.some(this.isScheduleableTask);
```

Comment:
any reason to change all these? They should all still work, shouldn't they? Typically we've used the ClassName.method() approach to call out static methods

Reply:
`this` is the only thorny part of the inheritance approach since we *must* invoke it with `this` if the overrideable errors are to work.

Rather than split between those that need it today and those that don't, I moved them all to `this` for consitency so a future implementer who might copy/paste doesn't accidentally use static :)

Different proposal though that helps solve this?